### ISSUE TITLE
minion crashed with empty grain

### ISSUE THREAD

**Issue by michaelforge** [saltstack/salt#23452]
Played with grains (append/delval). Then suddenly, minion became unavailable.

Was trying to `grains.delval` non-list grain, but didn't used `destructive=True`, so it became empty.
`service salt-minion restart` solved it, but I didn't expect minion to crash because ot this.

``` bash
root@salt-master:~# salt 'frontend*' grains.item roles
frontend2:
    ----------
    roles:
        frontend

root@salt-master:~# salt 'frontend*' grains.append roles debug
frontend2:
    The key roles is not a valid list

root@salt-master:~# salt 'frontend*' grains.delval roles
frontend2:
    None

root@salt-master:~# salt 'frontend*' grains.item roles
root@salt-master:~# salt 'frontend*' grains.append roles frontend
root@salt-master:~# salt 'frontend*' grains.item roles
root@salt-master:~# salt 'frontend*' grains.item roles
root@salt-master:~# salt 'front*' test.ping
root@salt-master:~# salt 'front*' test.ping -v
Executing job with jid 20150507133755164006
-------------------------------------------

frontend2:
    Minion did not return. [No response]
```

Looking at the minion:
``` bash
root@frontend2:~# salt-call status.master
[ERROR   ] Failed to read the virtual function for states: winrepo
Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/salt/loader.py", line 1072, in process_virtual
    virtual = mod.__virtual__()
  File "/usr/lib/python2.7/dist-packages/salt/states/winrepo.py", line 22, in __virtual__
    if 'salt-master' in __grains__.get('roles', [])
TypeError: argument of type 'NoneType' is not iterable
[WARNING ] /usr/lib/python2.7/dist-packages/salt/modules/config.py:258: DeprecationWarning: pillar_opts will default to False in the Lithium release
local:
    None
```

```
root@frontend2:~# salt-minion --versions-report
                  Salt: 2014.7.5
                Python: 2.7.6 (default, Mar 22 2014, 22:59:56)
                Jinja2: 2.7.2
              M2Crypto: 0.21.1
        msgpack-python: 0.3.0
          msgpack-pure: Not Installed
              pycrypto: 2.6.1
               libnacl: Not Installed
                PyYAML: 3.10
                 ioflo: Not Installed
                 PyZMQ: 14.0.1
                  RAET: Not Installed
                   ZMQ: 4.0.4
                  Mako: 0.9.1
 Debian source package: 2014.7.5+ds-1ubuntu1
```

**Comment by jfindlay** [comment]
@michaelforge, thanks for the report.

**Comment by cachedout** [comment]
Ouch! That's not a good result at all. :-/

I've fixed this up in #23639. Give it a shot and let me know how it works for you. Thanks!

**Comment by michaelforge** [comment]
@cachedout 
Thanks for your time!
Looks like our `salt-minion` upgraded yesterday to 2015.5.0, but I'll try anyway, your commit is still compatible. Will report back

**Comment by michaelforge** [comment]
@cachedout 
It's OK now! Can't reproduce this bug anymore.
But not sure what's exactly fixed it - your commit or upgrade to 2015.5.0. (; ͡°_ʖ ͡°)

**Comment by cachedout** [comment]
Sweet! I'll go ahead and close this issue then. Thanks for the help!
